### ISSUE TITLE
Should we use PowerShell directly in our builds?

### ISSUE THREAD

**Issue by CharliePoole** [TestCentric/testcentric-gui#36]
I'm making this a question for @TestCentric/gui-team 

The NUnit projects mostly used `build.cmd` as a front end to `build.ps1`.

Advantages:
 * Don't have to be at the PS prompt or type `powershell`
 * Don't have to type `./` before command name
 * More comfortable for old-timers like me!
 * Allows adding local options as a standard (see nunit/nunit build.cmd)

Disadvantages:
 * An extra command to check in.
 * Maybe confusing?

What do you guys think? How do you normally work? This doesn't impact users, just us.

**Comment by immeraufdemhund** [comment]
99% of the time if I'm using a terminal of sorts I'm using vscode with git bash as my default terminal. However.. if I click on a powershell file vscode automatically opens up a powershell prompt for me (a feature i hate). I personally prefer not using powershell however if i have to run some commands as admin I use powershell via the winkey+x keyboard shortcut on windows 10.

in summary. I dont like using it directly.

**Comment by mikkelbu** [comment]
Personally, I primarily use PowerShell - both at work and home - but it is not something I feel strongly about (during my studies I primarily used bash under Linux).

**Comment by ManfredLange** [comment]
In our projects we use Powershell terminals by default. Only if something else is more useful we switch. It took some time to get used to but at the moment we like Powershell as it does what we need for development, continuous integration and continuous deployment. Up until about 3 or 4 years ago we used a regular Windows batch file but realized that it was no longer sufficient. In particular we like that we can use any .NET library from within Powershell if needed -or- we can write cmdlets ourselves that behave like any other command you have available in Powershell.

We typically organize it like this: at the root of a repository we have a script named build.ps1 which is mostly empty except for a call to the actual build script ./tools/scripts/_build.ps1 which is launched in a separate Powershell instance as we tend to have one or more cmdlets in the code base so will be built and loaded by the build script before first use. When the second powershell script terminates the cmdlets are unloaded. This approach is not needed if you don't build cmdlets, don't use cmdlets or cmdlets don't change.

Having described our environment, here are my thoughts in respect to this project: I don't mind typing 'build' in a Powershell window then press the tab key which completes it to './build.ps1', then press enter. And usually that is one of the few commands we use so pressing the up key allows cycling through the history. Therefore most of the time it is up key plus enter key - two key presses. Not sure how to minimize this further.

I don't mind using build.cmd instead and/or having an additional file at the root of the repository that allows me building the system and run all the tests (which I prefer to be the case). I don't mind if we as a group would choose something other than Powershell. It's just a tool and as long as the tool does its job it can be anything. Please note that we rarely work on non-Windows platforms, so the picture (and our views) may change if that was the case. Others will have to contribute to this discussion from a non-Windows angle.

In summary: I'm not fuzzy. Either way would be fine with us.

**Comment by CharliePoole** [comment]
@ManfredLange I should have clarified a really big assumption, which was that we would have a cake build as NUnit has. Cake comes with build.ps1 and build.sh, which they call the bootstrap scripts. They automatically download nuget and cake itself before launching the build.cake script. @immeraufdemhund and @mikkelbu have worked more on the NUnit projects and so probably caught on to what I didn't explain. 😞 

Sounds like we are evenly divided about powerscript, so I pick both: we'll have build.ps1 as well as build.cmd, which invokes it.

Regarding the underlying scripts however... if anyone prefers something other than Cake, we could consider it. I have mixed feelings about it, primarily because they used to use a different compiler under Linux from Windows to compile the Cake (C#) script. That made things that worked in one environment sometimes fail in the other. I __think__ they are past that now, however. If someone wants to consider Rake, I'm willing to entertain it.

In any case, however, just to get things started, I'm going to push a Cake script so we at least have some kind of CI. As for powershell, I think my question is answered.

**Comment by CharliePoole** [comment]
BTW, I came across the idea of using the question label for a question by a team member to the rest of the team on another project. I hadn't done that before, using question primarily for user questions. Seems like a good idea!

**Comment by CharliePoole** [comment]
@ManfredLange BTW again... we have to use something that runs on Linux as well as Windows. Cake is C#, so it does that, but at the expense of having two separate bootstrap scripts, build.ps1 and build.sh. That's the part of it that seems like a bit of a pain.

**Comment by ManfredLange** [comment]
@CharliePoole I'm perfectly fine with using Cake. I understand and agree with the requirement to run cross-platform.

I haven't used it on Linux yet, but it appears as if [Powershell is available on Linux](https://github.com/PowerShell/PowerShell), too:.

Let's stick with Cake. It's been used in other NUnit related projects. Using Cake will make this project more consistent with those.

**Comment by CharliePoole** [comment]
Yeah... sticking with what we know and what works can't be all bad. 😄 

One of these days, I'd like to try Rake on a cross-platform project. Seems like it should cross over very nicely. Of course,

**Comment by immeraufdemhund** [comment]
We use rake at work. I'm not very fond of it though. I think that part of the problem is that there are still many windows specific things that it is doing and the way that it was architected wasn't very good. I've compared rake and cake and think that cake is much better for this purpose.

**Comment by CharliePoole** [comment]
Good to know!

**Comment by CharliePoole** [comment]
FYI, there's a PR for the initial build: AppVeyor only.
